File: cpp/Mod_stream.cpp
#pragma once

bool equals(List<Char> window, List<Char> pattern){
	for(int i=0; i<length(window); i++){
		if(get(pattern, i) == '*')
			continue;
		if(get(pattern, i) != get(window, i))
			return false;
	}
	return true;
}

List<Int> stream_debug(String str, String pat){
	cout << "string;pattern;window;result" << endl;
	List<Char> s = s2lc(str);
	List<Char> pattern = s2lc(pat);
	if(length(s) < length(pattern)){
		return List<Int>();
	}
	List<Char> window;
	List<Int> result;
	int i;
	for(i = 0; i<length(pattern); i++){
		cout << lc2s(s) << ";" << lc2s(pattern) << ";" << lc2s(window) << ";" << result << endl;
		if(empty(s)){
			return {};
		}
		Char f = first(s);
		window = append(window, f);
		s = removeFirst(s);
	}
	while(!empty(s)){
		cout << lc2s(s) << ";" << lc2s(pattern) << ";" << lc2s(window) << ";" << result << endl;
		if(equals(window, pattern)){
			result = append(result, i-length(pattern));
		}
		Char f = first(s);
		s = removeFirst(s);
		window = removeFirst(window);
		window = append(window, f);
		i++;
	}
	cout << lc2s(s) << ";" << lc2s(pattern) << ";" << lc2s(window) << ";" << result << endl;
	if(equals(window, pattern)){
		result = append(result, i-length(pattern));
	}
	return result;
}

List<Int> stream(String str, String pat){
	List<Char> s = s2lc(str);
	List<Char> pattern = s2lc(pat);
	if(length(s) < length(pattern)){
		return List<Int>();
	}
	List<Char> window;
	List<Int> result;
	int i;
    //Initialisierung
	for(i = 0; i<length(pattern); i++){
		if(empty(s)){
			return {};
		}
		Char f = first(s);
		window = append(window, f);
		s = removeFirst(s);
	}
	while(!empty(s)){
		if(equals(window, pattern)){
			result = append(result, i-length(pattern));
		}
		Char f = first(s);
		s = removeFirst(s);
		window = removeFirst(window);
		window = append(window, f);
		i++;
	}
	if(equals(window, pattern)){
		result = append(result, i-length(pattern));
	}
	return result;
}

File: hs/Mod_stream.hs
module Mod_stream ( stream ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types

stream :: String -> String -> [Int]
stream str pattern | length str < length pattern = []
                   | otherwise = trace "status;string;pattern;window" (streamHelper str pattern 0 [] [])
  
streamHelper :: String -> String -> Int -> String -> [Int] -> [Int]
streamHelper str pattern k window res 
 | k<length pattern = trace ("Aufbau;" ++ (lc2s str) ++ ";" ++ (lc2s pattern) ++ ";" ++ (lc2s window)) (streamHelper (removeFirst str) pattern (k+1) (append window (first str)) [])
 | empty str && equals window pattern = trace ("Ende+Match;" ++ (lc2s str) ++ ";" ++ (lc2s pattern) ++ ";" ++ (lc2s window)) (res ++ [k-(length pattern)])
 | empty str = trace ("Ende;" ++ (lc2s str) ++ ";" ++ (lc2s pattern) ++ ";" ++ (lc2s window) ) res
 | equals window pattern = trace ("Match;" ++ (lc2s str) ++ ";" ++ (lc2s pattern) ++ ";" ++ (lc2s window)) (streamHelper (removeFirst str) pattern (k+1) (shiftChar window (first str)) (res ++ [k-(length pattern)]))
 | otherwise = trace ("Kein Match;" ++ (lc2s str) ++ ";" ++ (lc2s pattern) ++ ";" ++ (lc2s window)) (streamHelper (removeFirst str) pattern (k+1) (shiftChar window (first str)) res)

equals window pattern 
 | empty window = True
 | first window == first pattern || first pattern == '*' = equals (removeFirst window) (removeFirst pattern)
 | otherwise = False

shiftChar :: String -> Char -> String
shiftChar str char = append (removeFirst str) char